cmd/kubeadm/app/cmd/join_test.go

```diff
@@ -243,3 +245,141 @@ func TestNewJoinData(t *testing.T) {
 		})
 	}
 }
+
+func initJoinOptions(token string, discovery kubeadmapiv1beta1.Discovery, controlPlane bool) *joinOptions {
+	joinOptions := newJoinOptions()
+	joinOptions.token = token
+	joinOptions.externalcfg.Discovery = discovery
+	joinOptions.controlPlane = controlPlane
+	return joinOptions
+}
+
+func TestAmendJoinOptions(t *testing.T) {
+	var tests = []struct {
+		description  string
+		args         []string
+		token        string
+		controlPlane bool
+		discovery    kubeadmapiv1beta1.Discovery
+		validate     func(*testing.T, *joinOptions)
+	}{
+		{
+			description:  "len(opt.token) > 0 and len(args) is 0 and controlPlane is false",
```

Comment:
We need to have smaller test units here. Tests like:
- `opt.token = []`
- `opt.token = [...]`
- `args = []`
- `args = ["single-api-server"]`
- `args = ["api-server1", "api-server2", ...]`
- `controlPlane = false`
- `controlPlane = true`
- ...
And thus, test things separately.